library-model/library-model-generator/src/main/java/com/uber/nullaway/libmodel/LibraryModelGenerator.java

```diff
@@ -0,0 +1,276 @@
+    if (absoluteDir.endsWith("/.")) {
+      absoluteDir = absoluteDir.substring(0, absoluteDir.length() - 2);
+    }
+    return Paths.get(absoluteDir);
+  }
+
+  private static class AnnotationCollectorCallback implements SourceRoot.Callback {
+
+    private final AnnotationCollectionVisitor annotationCollectionVisitor;
+
+    public AnnotationCollectorCallback(Map<String, MethodAnnotationsRecord> methodRecords) {
+      this.annotationCollectionVisitor = new AnnotationCollectionVisitor(methodRecords);
+    }
+
+    @Override
+    public Result process(Path localPath, Path absolutePath, ParseResult<CompilationUnit> result) {
+      Result res = Result.SAVE;
+      Optional<CompilationUnit> opt = result.getResult();
+      if (opt.isPresent()) {
+        CompilationUnit cu = opt.get();
+        cu.accept(annotationCollectionVisitor, null);
+      }
+      return res;
+    }
+  }
+
+  private static class AnnotationCollectionVisitor extends VoidVisitorAdapter<Void> {
+
+    private String packageName = "";
+    private boolean isJspecifyNullableImportPresent = false;
+    private boolean isNullMarked = false;
+    private Map<String, MethodAnnotationsRecord> methodRecords;
+    private static final String ARRAY_RETURN_TYPE_STRING = "Array";
+    private static final String NULL_MARKED = "NullMarked";
+    private static final String NULLABLE = "Nullable";
+    private static final String JSPECIFY_NULLABLE_IMPORT = "org.jspecify.annotations.Nullable";
+
+    public AnnotationCollectionVisitor(Map<String, MethodAnnotationsRecord> methodRecords) {
+      this.methodRecords = methodRecords;
+    }
+
+    @Override
+    public void visit(PackageDeclaration pd, Void arg) {
+      this.packageName = pd.getNameAsString();
+      super.visit(pd, null);
+    }
+
+    @Override
+    public void visit(ImportDeclaration id, Void arg) {
+      if (id.getName().toString().contains(JSPECIFY_NULLABLE_IMPORT)) {
+        this.isJspecifyNullableImportPresent = true;
+      }
+      super.visit(id, null);
+    }
+
+    @Override
+    public void visit(ClassOrInterfaceDeclaration cid, Void arg) {
+      /*Currently, this logic only supports @NullMarked annotations on top-level classes and
+      does not handle cases where @NullMarked appears on some nested classes but not others*/
+      cid.getAnnotations()
+          .forEach(
+              a -> {
+                if (a.getNameAsString().equalsIgnoreCase(NULL_MARKED)) {
+                  this.isNullMarked = true;
+                }
+              });
+      super.visit(cid, null);
+    }
+
+    @Override
+    public void visit(MethodDeclaration md, Void arg) {
+      Optional<Node> parentClassNode = md.getParentNode();
+      String parentClassName = "";
+      if (parentClassNode.isPresent()) {
+        if (parentClassNode.get() instanceof ClassOrInterfaceDeclaration) {
+          parentClassName = ((ClassOrInterfaceDeclaration) parentClassNode.get()).getNameAsString();
```

Comment:
This is just the simple name, right? So currently this breaks when dealing with nested/inner classes? i.e. the concatenation below:

```
            packageName
                + "."
                + parentClassName
                + ":"
                + getMethodReturnTypeString(md)
                [...]
```

Ends up producing `com.example.Bar:String baz()` for:

```
@NullMarked
class Foo {
   public class Bar {
      @Nullable public String baz() {...}
   }
}
```

correct?

Reply:
Yes, that is correct.

Reply:
@akulk022 is this easy to fix?  Or should it wait for a follow up?

Reply:
@msridhar I'm working on a fix since it seems like it should be easy enough.

Reply:
@akulk022 be sure to add a test for it as well

Reply:
@msridhar @lazaroclapp I've updated the logic to handle this scenario, I've also added a test for it. Does this work and are there more scenarios I should test?